com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -0,0 +1,412 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using MLAPI.Configuration;
+using MLAPI.Messaging;
+using MLAPI.NetworkVariable;
+using MLAPI.Serialization;
+using MLAPI.Serialization.Pooled;
+using MLAPI.Transports;
+using UnityEngine;
+using UnityEngine.UIElements;
+
+namespace MLAPI
+{
+    // Structure that acts as a key for a NetworkVariable
+    // Allows telling which variable we're talking about.
+    // Might include tick in a future milestone, to address past variable value
+    internal struct Key
+    {
+        public ulong NetworkObjectId; // the NetworkObjectId of the owning GameObject
+        public ushort BehaviourIndex; // the index of the behaviour in this GameObject
+        public ushort VariableIndex; // the index of the variable in this NetworkBehaviour
+    }
+
+    // Index for a NetworkVariable in our table of variables
+    // Store when a variable was written and where the variable is serialized
+    internal struct Entry
+    {
+        public Key Key;
+        public ushort TickWritten; // the network tick at which this variable was set
+        public ushort Position; // the offset in our Buffer
+        public ushort Length; // the length of the data in Buffer
+        public bool Fresh; // indicates entries that were just received
+
+        public const int NotFound = -1;
+    }
+
+    // A table of NetworkVariables that constitutes a Snapshot.
+    // Stores serialized NetworkVariables
+    // todo --M1--
+    // The EntryBlock will change for M1b with memory management, instead of just Beg, there will be data structure
+    // around available buffer, etc.
+    internal class EntryBlock
+    {
+        // todo --M1-- functionality to grow these will be needed in a later milestone
+        private const int k_MaxVariables = 64;
+        private const int k_BufferSize = 20000;
+
+        public byte[] Buffer = new byte[k_BufferSize];
+        public int Beg = 0; // todo: clarify usage. Right now, this is the beginning of the _free_ space.
+        public int End = 0;
+
+        public Entry[] Entries = new Entry[k_MaxVariables];
+        public int LastEntry = 0;
+        public MemoryStream Stream;
+
+        /// <summary>
+        /// Constructor
+        /// Allocated a MemoryStream to be reused for this EntryBlock
+        /// </summary>
+        public EntryBlock()
+        {
+            Stream = new MemoryStream(Buffer, 0, k_BufferSize);
```

Comment:
Is this something you would pool?  Don't we create a lot of EntryBlock's?  Or I wonder if this is one of the areas you're tuning in MS1b?  If so maybe note that long term we want to avoid doing these dynamic allocations for each snapshot?

Reply:
Actually, no, we don't create many of those. For M1, we have: (using new names)

```
public class SnapshotSystem 
{
        private Snapshot m_Snapshot = new Snapshot();
        private Snapshot m_ReceivedSnapshot = new Snapshot();
```
and that's it. One buffer to send and one to receive, and they get reused. Now, as we go to M1b and M2, we'll need 1 local + 1 per client, but allocations won't be frequent. I'll be willing to pool them if we get to MMO servers with frequent quits/joins. :-)